infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/replication/receive/StreamsSnapshotWriter.java

```diff
@@ -304,19 +304,25 @@ private void applyShadowStream(UUID streamId, long snapshot) {
         // as these streams will get trimmed and 'clear' will be a 'data loss'.
         if (MERGE_ONLY_STREAMS.contains(streamId)) {
             log.debug("Do not clear stream={} (merge stream)", streamId);
-        } else if (!replicatedStreamIds.contains(streamId)) {
-            log.trace("No data was written to stream {} on source or sink." +
-                "  Do not clear.", streamId);
-            return;
-        } else {
-            smrEntries.add(CLEAR_ENTRY);
         }
 
+        boolean shouldAddClearRecord = true;
         while (iterator.hasNext()) {
+            // append a clear record at the beginning of every non-merge-only streams
+            if(shouldAddClearRecord && !MERGE_ONLY_STREAMS.contains(streamId)) {
+                smrEntries.add(CLEAR_ENTRY);
+                shouldAddClearRecord = false;
+            }
+
```

Comment:
nit: I think the merge_only_streams check could be part of the initial value of shouldAddClearRecord?